strapi/strapi
# enhancement(database): add repair manager with component deletion fix

### What does it do?

- Adds a database repair manager 
- add a polymorph cleanup script for it
- call the repair function to clean up polymorph component_type when schema diff has changed

This PR is also serving as a proof-of-concept for the database repair function. Once decisions have been made about how it should work, I will add tests before merging, but I want it to stabilize first.

TODO: will add documentation in a later PR

### Why is it needed?

Cleans up orphaned data in deleted component tables on all environments without requiring a migration script

NOTE: anyone with an existing issue will still need to run `strapi.db.repair.removeOrphanTypes('component_type')` one time to fix it OR push a schema change to production to trigger this to run

### How to test it?

See https://github.com/strapi/strapi/issues/20931

- add components and dzs using components to your content types
- add data for those things
- delete the component used in the ct and dz
- everything should still work
- the content type _cmps table should not contain data for the deleted content type

There should not be any case where there is unintended data loss

### Related issue(s)/PR(s)

Alternative to: [v4 fix](https://github.com/strapi/strapi/pull/22207) [v5 fix](https://github.com/strapi/strapi/pull/22112)

Fixes https://github.com/strapi/strapi/issues/20931
